Introduce SSL_CTX_set_tls_cookie_{generate,verify}_cb

Ralith opened the pull request:
This documents the current semantics of the application cookie generate/verify callbacks used for TLS1.3 and DTLS stateless handshakes. Please review carefully for style, as I haven't done this before.

CC @mattcaswell

InfoHunter commented:
I guess there should be a 'RETURN VALUE' section in the doc. Refer to: https://github.com/openssl/openssl/pull/4976#issuecomment-354821733

InfoHunter commented on doc/man3/SSL_CTX_set_cookie_generate_cb.pod:
```diff
@@ -0,0 +1,56 @@
+
+=head1 SEE ALSO
+
+L<DTLSv1_listen(3)>, L<SSL_stateless(3)>,
```
No trailing comma at the end of line.

mattcaswell commented on doc/man3/SSL_CTX_set_cookie_generate_cb.pod:
```diff
@@ -0,0 +1,60 @@
+SSL_CTX_set_cookie_generate_cb() sets the callback used by L<DTLSv1_listen(3)>
+and L<SSL_stateless(3)> to generate the application-controlled portion of the
+cookie provided to clients in the HelloRetryRequest transmitted as a response to
+a ClientHello with a missing or invalid cookie. app_gen_cookie_cb() must write
```
The above describes SSL_stateless() which is a TLSv1.3 thing, but not so much DTLSv1_listen(). DLTSv1_listen() (which only supports <= DTLSv1.2), the callback provides the entire cookie (i.e. not just "the application-controlled portion"). Also it is created in response to a HelloVerifyRequest message in DTLSv1.2 and below.

mattcaswell commented on doc/man3/SSL_CTX_set_cookie_generate_cb.pod:
```diff
@@ -0,0 +1,60 @@
+ClientHello cookie is valid. A nonzero return value from app_verify_cookie_cb()
+communicates that the cookie is valid. The integrity of the entire cookie,
+including the application-controlled portion, is automatically verified using a
+HMAC before app_verify_cookie_cb() is called.
```
The HMAC thing only applies to SSL_stateless().

(As an aside I do wonder whether it would be wise to separate the DTLSv1_listen() cookie callback from the SSL_stateless() cookie callback - although superficially similar they are used differently which is kind of confusing).

richsalz commented:
One thing to do is run "util/find-doc-nits" and fix any errors it finds.  That will be a requirement before merging, anyway.

Ralith commented:
Per @mattcaswell's musing, I've drafted a dedicated pair of callbacks for `SSL_stateless` distinct from the `DTLSv1_listen` callbacks, and taken the opportunity to increase `SSL_COOKIE_LENGTH` somewhat arbitrarily to 4096. Input on better naming and cookie length very welcome, if this indeed seems like the way to go.

Am I correct to conclude that this is sufficiently nontrivial as to genuinely require a CLA?

richsalz commented:
Yeah, we'll need a CLA.

mattcaswell commented on doc/man3/SSL_CTX_set_tls_cookie_generate_cb.pod:
```diff
@@ -0,0 +1,60 @@
+determine whether the application-controlled portion of a ClientHello cookie is
+valid. A nonzero return value from app_verify_cookie_cb() communicates that the
+cookie is valid. The integrity of the entire cookie, including the
+application-controlled portion, is automatically verified using a HMAC before
```
Typo: probably should be "using HMAC"

Ralith replied:
I'm not sure about that. "using hash-based message authentication code" strikes me as incorrect because "code" is singular; it seems to me that it should be either "using a HMAC" or "using HMACs," but the latter would be incorrect because we're discussing a single abstract cookie here.

mattcaswell replied:
Well I would argue that you are using HMAC here to refer to the name of an algorithm, not a specific (singular) code that that algorithm generates. But I don't care enough about this to hold up the review, so I'll approve this PR.

Ralith replied:
Ah, I see where you're coming from. Tweaked the wording a little.

mattcaswell commented on include/openssl/ssl.h:
```diff
@@ -726,6 +726,19 @@ void SSL_CTX_set_cookie_verify_cb(SSL_CTX *ctx,
+                                        int (*gen_tls_cookie_cb) (SSL *ssl,
+                                                                  unsigned char
+                                                                  *cookie,
+                                                                  unsigned int
```
I think this should be a size_t

mattcaswell commented on include/openssl/ssl.h:
```diff
@@ -726,6 +726,19 @@ void SSL_CTX_set_cookie_verify_cb(SSL_CTX *ctx,
+                                      int (*verify_tls_cookie_cb) (SSL *ssl,
+                                                                   const unsigned char
+                                                                   *cookie,
+                                                                   unsigned int
```
Similarly here

mattcaswell commented:
Close/reopen to kick CLA bot

Ralith commented:
I made a trivial wording tweak and updated the `SSL_stateless` documentation to match the API changes.

mattcaswell commented:
Oh - actually one other thought springs to mind on this. At some point in the future we will want to add DTLS1.3 support, at which point the name `generate_tls_cookie_callback()` suddenly seems less appropriate. Maybe `generate_stateless_cookie_callback`?

Ralith commented:
To clarify, are you suggesting that for DTLS1.3 the callbacks introduced in this PR will be reused, and should therefore be named more generically?

mattcaswell commented:
> To clarify, are you suggesting that for DTLS1.3 the callbacks introduced in this PR will be reused, and should therefore be named more generically?

Yes

Ralith commented:
Drafted, along with some doc fixes. Looks like the longest symbol in OpenSSL yet! Sacrifices some clarity regarding the different from the old functions, but I guess that's a lost cause if we want them to be reused in the future anyway.

mattcaswell commented:
Ping @openssl/committers for second review.

kaduk commented on apps/s_cb.c:
```diff
@@ -752,6 +752,22 @@ int verify_cookie_callback(SSL *ssl, const unsigned char *cookie,
+int verify_stateless_cookie_callback(SSL *ssl, const unsigned char *cookie,
+                                     size_t cookie_len)
+{
+    return verify_cookie_callback(ssl, cookie, cookie_len);
```
Why do we not need to check that ```cookie_len <= UINT_MAX``` to match the ```verify_cookie_callback()``` prototype?

Ralith replied:
A TLS1.3 cookie is at most 2^16-1 bytes. Is that not a sufficiently strong bound?

kaduk replied:
Probably, assuming the state machine checks that before calling into us (which I'm sure it does because it uses the PACKET functions).

kaduk commented on doc/man3/SSL_CTX_set_stateless_cookie_generate_cb.pod:
```diff
@@ -0,0 +1,62 @@
+                                  const unsigned char
+                                  *cookie,
+                                  size_t
+                                  cookie_len));
```
How did we choose the indentation level for these arguments?  (It doesn't look wrong per se, it just seems pretty arbitrary.)

Ralith replied:
Fell out of alignment when the argument was renamed. Thanks for the catch, fixing.

kaduk commented on ssl/ssl_sess.c:
```diff
@@ -1269,4 +1269,22 @@ void SSL_CTX_set_cookie_verify_cb(SSL_CTX *ctx,
+    SSL_CTX *ctx,
+    int (*cb) (SSL *ssl,
+               const unsigned char *cookie,
+               size_t cookie_len))
```
Did we consider named typedefs for these callback functions?

Ralith replied:
My intention was to imitate the style of the surrounding code in ssl.h.

kaduk commented on test/sslapitest.c:
```diff
@@ -2583,6 +2583,21 @@ static int verify_cookie_callback(SSL *ssl, const unsigned char *cookie,
+static int verify_stateless_cookie_callback(SSL *ssl, const unsigned char *cookie,
+                                      size_t cookie_len)
+{
+    return verify_cookie_callback(ssl, cookie, cookie_len);
```
(Same as for s_cb.c)

kaduk commented on include/openssl/ssl.h:
```diff
@@ -546,8 +546,8 @@ typedef int (*SSL_verify_cb)(int preverify_ok, X509_STORE_CTX *x509_ctx);
-/* Length of a TLSv1.3 cookie */
-# define SSL_COOKIE_LENGTH                       255
+/* Maximum length of the application-controlled segment of a a TLSv1.3 cookie */
+# define SSL_COOKIE_LENGTH                       4096
```
For some reason I am remembering discussion that changing a defined value in a public header like this would be considered an ABI break; am I misremembering?

mattcaswell replied:
This is a new define in 1.1.1 so that doesn't apply here.

Ralith commented:
Travis failures look spurious.

Ralith commented:
Rebased. @kaduk, have I addressed your comments to your satisfaction? I'm happy to add typedefs and/or paranoid bounds-checking if we really want them.

mattcaswell commented:
Reconfirm.

mattcaswell commented:
Pushed! Thanks.
